File: javascript_code/line/algo/sort/InsertionSorter.js


// 时间复杂度：O(n^2)
// 空间复杂度：O(1)
const sort = function (data) {
    if (data == null || data.length == 0) return
    // 插入排序的轮数
    for (let i = 1; i < data.length; i++) {
        const tmp = data[i]
        let j;
        for (j = i; j > 0; j--) {
            if (tmp < data[j - 1]) {
                // 将较大的元素总是向右移动
                data[j] = data[j - 1]
            } else {
                break
            }
        }
        // 找到 i 对应的元素需要插入的位置
        data[j] = tmp
    }
}

// 时间复杂度：O(n^2)
// 空间复杂度：O(1)
const sort1 = function (data) {
    if (data == null || data.length == 0) return
    // 插入排序的轮数
    for (let i = 1; i < data.length; i++) {
        for (let j = i; j > 0; j--) {
            if (data[j] < data[j - 1]) {
                [data[j], data[j - 1]] = [data[j - 1], data[j]]
            } else {
                break
            }
        }
    }
}

data = [12, 23, 36, 9, 24, 42]
sort(data)
console.log(data)

File: python_code/line/algo/sort/InsertionSorter.py


# 时间复杂度：O(n^2)
# 间复杂度：O(1)
def sort(data):
    if data is None or len(data) <= 1:
        return
    # 插入排序的轮数
    for i in range(1, len(data)):
        tmp = data[i]
        j = i
        while j > 0:
            if tmp < data[j - 1]:
                # 将较大的元素总是向右移动
                data[j] = data[j - 1]
            else:
                break
            j -= 1
        # 找到 i 对应的元素需要插入的位置
        data[j] = tmp


# 时间复杂度：O(n^2)
# 间复杂度：O(1)
def sort1(data):
    if data is None or len(data) <= 1:
        return
    # 插入排序的轮数
    for i in range(1, len(data)):
        tmp = data[i]
        for j in range(i, 0, -1):
            if tmp < data[j - 1]:
                data[j], data[j - 1] = data[j - 1], data[j]
            else:
                break


if __name__ == '__main__':
    data = [12, 23, 36, 9, 24, 42]
    sort1(data)
    print(data)
